client/db/src/pinned_blocks_cache.rs

```diff
@@ -0,0 +1,207 @@
+// This file is part of Substrate.
+
+// Copyright (C) 2017-2023 Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+use crate::LOG_TARGET;
+use schnellru::{Limiter, LruMap};
+use sp_runtime::{traits::Block as BlockT, Justifications};
+
+const PINNING_CACHE_SIZE: u32 = 1024;
+
+struct PinnedBlockCacheEntry<Block: BlockT> {
+	ref_count: u32,
+	pub justifications: Option<Option<Justifications>>,
+	pub body: Option<Option<Vec<Block::Extrinsic>>>,
+}
+
+impl<Block: BlockT> Default for PinnedBlockCacheEntry<Block> {
+	fn default() -> Self {
+		Self { ref_count: 0, justifications: None, body: None }
+	}
+}
+
+impl<Block: BlockT> PinnedBlockCacheEntry<Block> {
+	pub fn decrease_ref(&mut self) {
+		self.ref_count = self.ref_count.saturating_sub(1);
+	}
+
+	pub fn increase_ref(&mut self) {
+		self.ref_count = self.ref_count.saturating_add(1);
+	}
+
+	pub fn has_no_references(&self) -> bool {
+		self.ref_count == 0
+	}
+}
+
+/// A limiter for a map which is limited by the number of elements.
+#[derive(Copy, Clone, Debug)]
+struct LoggingByLength {
+	max_length: u32,
```

Comment:
It's pasted from `schnellru`, and there it's `u32` because the limiter uses `u32` as the type for the `prev`/`next` links between entries in the map (the `LinkType`), and the choice of the link type naturally limits the maximum number of elements that the map can hold (if you can only index `u32::MAX` items then you can never hold more than `u32::MAX` items, hence the `max_length` is of the same type to reflect that).

Reply:
But why is then this: https://github.com/paritytech/substrate/pull/13157/files/81f749368c52775666acaae5997dd69729ccce28#diff-a1cf47ee5e082481ee054b73b6ec409e6dc3590c65706e45bfccd44bdc5e73baR72 using usize? :D (That was the reason I added this comment)

Reply:
Hm, I guess, good point, that could also be `Self::LinkType` instead of `usize` since it's guaranteed to never exceed it. :P